sphinx_immaterial/code_annotations.py

```diff
@@ -0,0 +1,79 @@
+from typing import List
+from docutils import nodes
+from sphinx.application import Sphinx
+from sphinx.util.docutils import SphinxDirective
+from sphinx.writers.html import HTMLTranslator
+from sphinx.util.logging import getLogger
+from . import html_translator_mixin
+
+LOGGER = getLogger(__name__)
+
+
+class annotations_list(nodes.container):
+    pass
+
+
+class CodeAnnotations(SphinxDirective):
+    """Take a list of annotations and move them into a previous node's parent."""
+
+    has_content = True
+
+    def run(self) -> List[nodes.Node]:
+        """Run the directive."""
+        self.assert_has_content()
+
+        div = annotations_list("")
+        self.state.nested_parse(self.content, self.content_offset, div)
+        self.set_source_info(div)
+        if not isinstance(div.children[0], nodes.enumerated_list):
+            LOGGER.error(
+                "The code-annotations directive only accepts an enumerated list.",
+                location=div,
+            )
+        return [div]
+
+
+def visit_annotations_list(self: HTMLTranslator, node: annotations_list):
+    # the content should have been moved into the literal block before calling this
+    raise nodes.SkipNode()  # see visit_literal_block() override
+
+
+@html_translator_mixin.override
+def visit_literal_block(
+    self: html_translator_mixin.HTMLTranslatorMixin,
+    node: nodes.literal_block,
+    super_func: html_translator_mixin.BaseVisitCallback[nodes.literal_block],
+):
+    #  we need to be sure that the annotated_list is the next node
+    next_element = node.next_node(nodes.Element, siblings=True, descend=False)
+    parent = node.parent
+    if isinstance(parent.children[0], nodes.caption):
+        # literal blocks with a caption have added parent
+        parent = parent.parent
+        next_element = node.parent.next_node(
+            nodes.Element, siblings=True, descend=False
+        )
+    if isinstance(next_element, annotations_list):
+        try:
+            super_func(self, node)
+        except nodes.SkipNode:
+            pass  # we're not done yet
+        # Add annotate class to the highlighted div tag (generated by pygments), and
+        # remove the closing div tag (`[:-7]`) of the highlighted div's parent node.
+        self.body[-1] = self.body[-1][:-7].replace(
```

Comment:
add comment (to clarify to reader) and assert here regarding content of `self.body[-1]` as check against incompatible changes to sphinx

Reply:
see changes in https://github.com/jbms/sphinx-immaterial/pull/176/commits/201f719be816467188da693a56168196260c67ea

I hope this satisfies the request

Reply:
Thanks